### PR TITLE
FlipViewExtensions implementation

### PR DESCRIPTION
**Pull request by pictos**
GitHub Issue (If applicable): #360

## PR Type

What kind of change does this PR introduce?
- Feature

## What is the current behavior?
- Devs sad because they need to set a lot of things in order to link the `FlipView` and `PipsPager` control;
- Devs sad because there's no easy way to navigate throw `FlipView` items using external `Button`s

## What is the new behavior?
- Devs happy because now it's easier to link the `FlipView` and `PipsPager` control;
- Devs happy because they can easily navigate throw `FlipView` items using external `Button`s

## PR Checklist
Please check if your PR fulfills the following requirements:
- [ ] Tested code with current [supported SDKs](../README.md#supported)
- [ ] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [ ] Associated with an issue (GitHub or internal)

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by kazo0** [comment]
We should be including runtime/UI Tests for most changes moving forward.

Are we able to squeeze those into this PR or does this need to merge ASAP? Otherwise, we can pair on writing the tests after the release

File: doc/helpers/FlipView-extensions.md [hunk 1]
```diff
@@ -0,0 +1,43 @@
+# SelectorExtensions Attached Properties
+
+Provides additional features for `FlipView`.
+
+## Properties
+Property|Type|Description
+-|-|-
+Next|FlipView| Sets the `FlipView` to will be controlled.
```
Comment (kazo0) [inline doc/helpers/FlipView-extensions.md:8]:
```suggestion
Next|FlipView| Sets the `FlipView` that should be moved to its next item
```

File: doc/helpers/FlipView-extensions.md [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+Previous|FlipView| Sets the `FlipView` that will controled.
```
Comment (kazo0) [inline doc/helpers/FlipView-extensions.md:9]:
```suggestion
Previous|FlipView| Sets the `FlipView` that should be moved to its previous item
```

**Review by kazo0 — APPROVED** [review]

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+	  Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"
+	  mc:Ignorable="d">
+
+	<sample:SamplePageLayout x:Name="SamplePageLayout" IsDesignAgnostic="True">
+		<sample:SamplePageLayout.DesignAgnosticTemplate>
+			<DataTemplate>
+				<StackPanel Margin="0,20" Spacing="20">
+					<StackPanel.Resources>
+						<ResourceDictionary>
+							<Style TargetType="x:TextBlock">
```
Comment (kazo0) [inline samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml:19]:
```suggestion
							<Style TargetType="TextBlock">
```

**Comment by jeromelaban** [comment]
Let's take the time to add a runtime test as well.

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+					<StackPanel.Resources>
+						<ResourceDictionary>
+							<Style TargetType="TextBlock">
+								<Setter Property="HorizontalAlignment" Value="Center" />
+								<Setter Property="TextWrapping" Value="Wrap" />
+							</Style>
+						</ResourceDictionary>
+					</StackPanel.Resources>
+
+					<TextBlock Text="Try to navigate throw the FlipView using the buttons inside it. " />
```
Comment (Xiaoy312) [inline samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml:26]:
```suggestion
					<TextBlock Text="Try to navigate the FlipView using the buttons inside it. " />
```
Or, did you mean "... using swipe gesture or the buttons inside it."?

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+					<msui:PipsPager x:Name="pipsPager"
+									MinHeight="100"
+									HorizontalAlignment="Center"
+									MaxVisiblePips="10" />
+					<TextBlock Text="To add a new view, press the button bellow" />
+					<Button x:Name="AddNewPageButton"
+							HorizontalAlignment="Center"
+							Content="Add" />
+					<Button x:Name="DebugVTButton" Content="DebugVT" />
```
Comment (Xiaoy312) [inline samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml:87]:
```suggestion
```
debug code to remove

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+					grid.Children.Add(bt2);
+
+					flipView.Items.Add(grid);
+				};
+
+				var debugVT = this.SamplePageLayout.GetSampleChild<Button>(Design.Agnostic, "DebugVTButton");
+				debugVT.Click += (s, e) =>
+				{
+					var tree = flipView.TreeGraph();
+				};
```
Comment (Xiaoy312) [inline samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml.cs:92]:
```suggestion
```
debug code to remove

File: src/Uno.Toolkit.UI/Behaviors/FlipView.xaml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿<ResourceDictionary xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation" xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">
+	<Style x:Key="FlipViewNoArrowsStyle" TargetType="FlipView">
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/FlipView.xaml:2]:
```suggestion
	<Style x:Key="NoArrowsFlipViewStyle" TargetType="FlipView">
```
Typically the convention is "[qualifiers...}{control-type-name}Style", with few exceptions (eg: textblock styles that needs to match material design).

File: src/Uno.Toolkit.UI/Behaviors/SelectorExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+{
+	private readonly static WeakAttachedDictionary<Selector, string> events = new();
+	private const string SelectorAttachedPropertyKey = "Selector";
+
+	/// <summary>
+	/// Backing property for the <see cref="PipsPager"/> that will be linked to the desired <see cref="Selector"/> control.
+	/// </summary>
+	#region DependencyProperty: PipsPager
+	public static DependencyProperty PipsPagerProperty { get; } =
+	DependencyProperty.RegisterAttached("PipsPager", typeof(PipsPager), typeof(SelectorExtensions), new PropertyMetadata(null, OnPipsPagerChanged));
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/SelectorExtensions.cs:47]: indentation
```suggestion
	public static DependencyProperty PipsPagerProperty { get; } =
		DependencyProperty.RegisterAttached("PipsPager", typeof(PipsPager), typeof(SelectorExtensions), new PropertyMetadata(null, OnPipsPagerChanged));
```

File: src/Uno.Toolkit.UI/Behaviors/FlipViewExtensions.cs [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+
+namespace Uno.Toolkit.UI;
+/// <summary>
+/// Extensions for <see cref="FlipView"/>
+/// </summary>
+public static class FlipViewExtensions
+{
+	#region DependencyProperty: Next
+	public static DependencyProperty NextProperty { get; } =
+	DependencyProperty.RegisterAttached("Next", typeof(FlipView), typeof(FlipViewExtensions), new PropertyMetadata(null, OnNextChanged));
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/FlipViewExtensions.cs:33]:
```suggestion
		DependencyProperty.RegisterAttached("Next", typeof(FlipView), typeof(FlipViewExtensions), new PropertyMetadata(null, OnNextChanged));
```

File: src/Uno.Toolkit.UI/Behaviors/FlipViewExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+
+	public static void SetNext(Button element, FlipView value) => element.SetValue(NextProperty, value);
+
+	public static FlipView? GetNext(Button element) => (FlipView?)element.GetValue(NextProperty);
+
+	#endregion
+	#region DependencyProperty: Previous
+	public static DependencyProperty PreviousProperty { get; } =
+	DependencyProperty.RegisterAttached("Previous", typeof(FlipView), typeof(FlipViewExtensions), new PropertyMetadata(null, OnPreviousChanged));
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/FlipViewExtensions.cs:42]:
```suggestion
		DependencyProperty.RegisterAttached("Previous", typeof(FlipView), typeof(FlipViewExtensions), new PropertyMetadata(null, OnPreviousChanged));
```

File: src/Uno.Toolkit.UI/Behaviors/SelectorExtensions.cs [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+	{
+		var value = events.GetValue<VectorChangedEventHandler<object>>(selector, SelectorAttachedPropertyKey);
+
+		if (value is not null)
+		{
+			selector.Items.VectorChanged -= value;
+		}
+	}
+
+	#region SelectionOffset Attached Property
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UI/Behaviors/SelectorExtensions.cs:129]: pick a lane:
```
#region PipsPager Dependency Property
#region SelectionOffset Attached Property // chosen

#region DependencyProperty: PipsPager // chosen
#region AttachedProperty: SelectionOffset
```

File: src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_FlipViewExtensions.cs [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+
+		var oldFlipViewSelectedIndex = flipView.GetDependencyPropertyValue<int>("SelectedIndex");
+		Assert.AreEqual(0, oldFlipViewSelectedIndex);
+
+		App.Tap("BtnNext1");
+		App.Tap("BtnNext2");
+
+		var newFlipViewSelectedIndex = flipView.GetDependencyPropertyValue<int>("SelectedIndex");
+
+		Assert.AreNotEqual(oldFlipViewSelectedIndex, newFlipViewSelectedIndex);
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_FlipViewExtensions.cs:23]: redundant, you will check against 2 anyways
```suggestion
```

File: src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_FlipViewExtensions.cs [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+	(QueryEx flipView, QueryEx pipsPager) GetPipsPagerAndFlipViewElements()
+	{
+		var flipView = App.MarkedAnywhere("flipView");
+		var pipsPager = App.MarkedAnywhere("pipsPager");
+
+		Assert.IsNotNull(flipView);
+		Assert.IsNotNull(pipsPager);
+
+		return (flipView, pipsPager);
+	}
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_FlipViewExtensions.cs:52]: unused copy-pasta
```suggestion

	(QueryEx flipView, QueryEx pipsPager) GetPipsPagerAndFlipViewElements()
	{
		var flipView = App.MarkedAnywhere("flipView");
		var pipsPager = App.MarkedAnywhere("pipsPager");

		Assert.IsNotNull(flipView);
		Assert.IsNotNull(pipsPager);

		return (flipView, pipsPager);
	}
```

File: src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+	public void Number_Of_Items_Should_Be_The_Same_AfterAddingMoreItems()
+	{
+		var (flipViewCounter, pipsPager) = GetPipsPagerAndLabelElements();
+
+		App.Tap("AddNewPageButton");
+
+		var pipsPagerItemsCount = pipsPager.GetDependencyPropertyValue<int>("NumberOfPages");
+		var flipViewCount = flipViewCounter.GetDependencyPropertyValue<string>("Text");
+
+		Assert.AreEqual(pipsPagerItemsCount, int.Parse(flipViewCount));
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs:32]: This test fails to validate if the count has actually increased after triggering `AddNewPageButton`.
You should check the before AND after for both `NumberOfPages` + `Text`.

File: src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+	{
+		var (flipView, pipsPager) = GetPipsPagerAndFlipViewElements();
+
+		App.Tap("BtnNext1");
+		App.Tap("BtnNext2");
+
+		var pipsPagerSelectedIndex = pipsPager.GetDependencyPropertyValue<int>("SelectedPageIndex");
+		var flipViewSelectedIndex = flipView.GetDependencyPropertyValue<int>("SelectedIndex");
+
+		Assert.AreEqual(pipsPagerSelectedIndex, flipViewSelectedIndex);
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs:46]: same here "before AND after"

File: src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+	}
+
+	(QueryEx flipViewCounter, QueryEx pipsPager) GetPipsPagerAndLabelElements()
```
Comment (Xiaoy312) [inline src/Uno.Toolkit.UITest/Behaviors/FlipViewPipsPagerBehavior/Given_PipsPagerExtensions.cs:49]: These two helper methods don't bring much to the table.
Assert.IsNotNull vs NullReferenceException isnt that much different in term of context provided.

File: samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml (hunk 4 above)
Comment (pictos) [reply samples/Uno.Toolkit.Samples/Uno.Toolkit.Samples.Shared/Content/Controls/FlipViewSamplePage.xaml:26]: The buttons inside it

File: doc/helpers/FlipView-extensions.md [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+                    utu:FlipViewExtensions.Next="{Binding ElementName=flipView}"
+                    Content="Next" />
+            </Grid>
+        </FlipView.Items>
+    </FlipView>
+```
+
+## Customization
+
+The `FlipView` control for desktop targets will present arrow buttons to allow the user to click on the to navigate between the content inside the `FlipView`, if you don't want these arrows to be shown, you can add our `Style` to remove them. Your xaml code will be.
```
Comment (kazo0) [inline doc/helpers/FlipView-extensions.md:48]:
```suggestion
The `FlipView` control for desktop targets will present arrow buttons to allow the user to click on the to navigate between the content inside the `FlipView`, if you don't want these arrows to be shown, you can use the `FlipViewNoArrowsStyle` as shown in the example below:
```

File: doc/helpers/Selector-extensions.md [hunk 18]
```diff
@@ -0,0 +1,26 @@
+# SelectorExtensions Attached Properties
+
+Provides an easy integration between `PipsPager` and `Selector` controls.
+
+## Properties
+Property|Type|Description
+-|-|-
+PipsPagerProperty|PipsPager| Backing property for the `PipsPager` that will interact with the `Selector` control.
+
+When the `SelectorExtensions.PipsPager` is set, the control will take care of updating the number of items and the `SelectedIndex` automatically.
```
Comment (kazo0) [inline doc/helpers/Selector-extensions.md:10]:
```suggestion
When the `SelectorExtensions.PipsPager` is set, the control will take care of updating the `NumberOfPages` and `SelectedIndex` properties automatically.
```

File: doc/helpers/Selector-extensions.md [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+## Usage
+```xml
+xmlns:utu="using:Uno.Toolkit.UI"
+...
+
+<Grid>
+    <FlipView
+        utu:SelectorExtensions.PipsPager="{Binding ElementName=pipsPager}"
+        Background="Blue">
+    <msui:PipsPager
```
Comment (kazo0) [inline doc/helpers/Selector-extensions.md:21]:
```suggestion
    <muxc:PipsPager
```

File: doc/helpers/Selector-extensions.md [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+## Properties
+Property|Type|Description
+-|-|-
+PipsPagerProperty|PipsPager| Backing property for the `PipsPager` that will interact with the `Selector` control.
+
+When the `SelectorExtensions.PipsPager` is set, the control will take care of updating the number of items and the `SelectedIndex` automatically.
+
+## Usage
+```xml
+xmlns:utu="using:Uno.Toolkit.UI"
```
Comment (kazo0) [inline doc/helpers/Selector-extensions.md:14]:
```suggestion
xmlns:utu="using:Uno.Toolkit.UI"
xmlns:muxc="using:Microsoft.UI.Xaml.Controls"
```

**Review by kazo0 — APPROVED** [review]

**Comment by Xiaoy312** [comment]
There are multiple comments that are marked as resolved, but aren't really addressed.

**Pull request merged**